lib/THC/THCApply.cuh

```diff
@@ -164,55 +171,60 @@ bool THCudaTensor_pointwiseApply1(THCState* state,
   // (or vice versa), the contiguous tensor can be collapsed to one
   // dimension, and the loop to translate the linear index to the array
   // index can be similarly collapsed. That is what this unrolling is for.
-#define HANDLE_CASE(TYPE, A)                                   \
-  THCudaTensor_pointwiseApply1<Op, TYPE, A>                    \
-    <<<grid, block, 0, THCState_getCurrentStream(state)>>>(    \
+#define HANDLE_CASE(TYPE, A)                                            \
+  kernelPointwiseApply1<Op,                                             \
+                        typename TensorUtils<TensorTypeA>::DataType,   \
+                        TYPE, A>                                        \
+    <<<grid, block, 0, THCState_getCurrentStream(state)>>>(             \
       aInfo, (TYPE) totalElements, op);
 
-#define HANDLE_A_CASE(TYPE, A)                      \
-  {                                                 \
-    if (aInfo.isContiguous()) {                     \
-      HANDLE_CASE(TYPE, -2);                        \
-    } else {                                        \
-      switch (A) {                                  \
-        case 1:                                     \
-          HANDLE_CASE(TYPE, 1);                     \
-          break;                                    \
-        case 2:                                     \
-          HANDLE_CASE(TYPE, 2);                     \
-          break;                                    \
-        case 3:                                     \
-          HANDLE_CASE(TYPE, 3);                     \
-          break;                                    \
-        default:                                    \
-          HANDLE_CASE(TYPE, -1);                    \
-          break;                                    \
-      }                                             \
-    }                                               \
+#define HANDLE_A_CASE(TYPE, A)                  \
+  {                                             \
+    if (aInfo.isContiguous()) {                 \
+      HANDLE_CASE(TYPE, -2);                    \
+    } else {                                    \
+      switch (A) {                              \
+        case 1:                                 \
+        HANDLE_CASE(TYPE, 1);                   \
+        break;                                  \
+        case 2:                                 \
+        HANDLE_CASE(TYPE, 2);                   \
+        break;                                  \
+        default:                                \
+        HANDLE_CASE(TYPE, -1);                  \
+        break;                                  \
```

Comment:
I deleted the HANDLE_CASE(TYPE, 3) specializations from THCApply.cuh in an attempt to cut down on compile time (should address some of @hughperkins issues)